Source/Dafny/Verifier/Translator.BoogieFactory.cs

```diff
@@ -867,5 +867,25 @@ enum BuiltinFunction {
       fvars.Add(BplFormalVar(name, ty, incoming, out var e));
       return e;
     }
+
+    public static IToken ToDafnyToken(Bpl.IToken exprTok) {
```

Comment:
When do we need to go from Boogie to Dafny tokens? For translating I would expect us to only go from Dafny to Boogie tokens.

Reply:
Because we have Dafny expressions that wrap around Boogie expressions in multiple places. And they use to copy the token of the wrapped expressions.

Reply:
> Because we have Dafny expressions that wrap around Boogie expressions in multiple places.

Could you give me a pointer for where to find this?

Reply:
* `Bpl.PredicateCmd AssertNS` in translator.cs uses two tokens, one which is the expression token and the other one the refinement token. The check uses `RefinementToken.IsInherited`  which is defined in the resolver pipeline and requires a Dafny token.
https://github.com/dafny-lang/dafny/blob/07d614f193502b07b4cea671c2eeb2018ced726d/Source/Dafny/Verifier/Translator.cs#L8413-L8419
* We first convert an expression to boogie induction cases (it returns a sequence of `Boogie.Expr`), and then use their tokens to build a `NestedToken`, which is requires Dafny tokens as inputs.
https://github.com/dafny-lang/dafny/blob/07d614f193502b07b4cea671c2eeb2018ced726d/Source/Dafny/Verifier/Translator.cs#L11416-L11419
* When splitting an expression, we create multiple `Boogie.Expr`, and then we recreate other NestedTokens based on their tokens.
https://github.com/dafny-lang/dafny/blob/07d614f193502b07b4cea671c2eeb2018ced726d/Source/Dafny/Verifier/Translator.TrStatement.cs#L99-L104
* When substituting expressions, we use a BoogieWrapper that takes a Boogie expression and converts it into a Dafny expression, and there are 27 such calls to `new BoogieWrapper`:
https://github.com/dafny-lang/dafny/blob/master/Source/Dafny/Verifier/Translator.cs#L10633-L10638
* Also, when recovering counter-examples and Boogie expressions after verification, we get Boogie tokens, but now the LSP uses Dafny tokens.